Automattic/wp-calypso
# Port close-on-esc mixin

### Summary

This PR is a slightly spruced up & more complete version of #10268 (I've opened a new PR on an internal branch vs a fork).

You'll find some details and discussion over at #10268, and more up to date details below.

### Motivation

I'm hoping that the dependancy on mixins can be completely removed by replacing their usage with components. That would allow for the usage of `React.createClass` to be fully replaced with ES6 syntax & `extends component` / `extends pureComponent` and get us a step closer to React v16.

### Testing

To test these changes I've been doing the following:

- Go to a blog post
- Select 'Featured Image' > 'Set Featured Image' from the side-bar
  <img width="228" alt="featured-image" src="https://cloud.githubusercontent.com/assets/4335450/21477448/20ec2d3a-cb96-11e6-9094-50eb2c8dba78.png">
- Upload or select an image from your media library
- follow through to the image editor for that image
  ![image-editor](https://cloud.githubusercontent.com/assets/4335450/21477547/f941f87c-cb96-11e6-896d-221121b5ce22.gif)
- There's now a 'stack' of escapable components, `dialog-base` and `image-editor`
- Press esc once, the image editor should close.
- Press esc again, the media library (uses dialog-base) should now close too.
- Repeat this process but hold 'esc' down.
- Both components should close in succession.

cc'ing for continuation of #10268: @dmsnell, @gwwar, @gziolo, @retrofox, @ockham - FFTI though.

(I may still need to give https://github.com/Automattic/wp-calypso/pull/10268#issuecomment-275473195 and some other points some attention)
